### PR TITLE
[python] Add type hints to helpers/parameter_generator.py

### PR DESCRIPTION
**Pull request by sagnik1511**
contributes to #3756

Added type hints on helpers/parameter_generator.py

### CODE REVIEW

File: helpers/parameter_generator.py [hunk 1]
```diff
@@ -99,7 +104,9 @@ def get_names(infos):
     return names
 
 
-def get_alias(infos):
+def get_alias(
+    infos: List
+) -> List[Tuple]:
```
Comment (jameslamb) [inline helpers/parameter_generator.py:109]:
```suggestion
) -> List[Tuple[str, str]]:
```

Please be specific about the types of values that will be found in `Dict`, `List`, and `Tuple`  objects. This allows static analysis tools like `mypy` to find more issues.

File: helpers/parameter_generator.py [hunk 2]
```diff
@@ -7,9 +7,12 @@
 from the information in LightGBM/include/LightGBM/config.h file.
 """
 from pathlib import Path
+from typing import List, Optional, Tuple
 
 
-def get_parameter_infos(config_hpp):
+def get_parameter_infos(
+    config_hpp: Path
+) -> Tuple[List, List]:
```
Comment (jameslamb) [inline helpers/parameter_generator.py:15]:
```suggestion
) -> Tuple[List[Tuple[str, int]], List[Dict[str, Any]]]:
```

File: helpers/parameter_generator.py [hunk 3]
```diff
@@ -123,7 +130,10 @@ def get_alias(infos):
     return pairs
 
 
-def parse_check(check, reverse=False):
+def parse_check(
+    check: Optional[str],
```
Comment (jameslamb) [inline helpers/parameter_generator.py:134]:
```suggestion
    check: str,
```

The type `Optional` means that a variable can take the value `None`. You can see that in the `typing` documentation, https://docs.python.org/3/library/typing.html#typing.Union.

> You can use `Optional[X]` as a shorthand for `Union[X, None]`.

`check` in this function cannot take the value `None`, so `Optional` should be removed.

File: helpers/parameter_generator.py [hunk 4]
```diff
@@ -123,7 +130,10 @@ def get_alias(infos):
...
+    reverse: bool = False
+) -> Tuple[Optional[str], Optional[str]]:
```
Comment (jameslamb) [inline helpers/parameter_generator.py:136]:
```suggestion
) -> Tuple[str, str]:
```

File: helpers/parameter_generator.py [hunk 5]
```diff
@@ -151,7 +161,11 @@ def parse_check(check, reverse=False):
         return check[idx:], check[:idx]
 
 
-def set_one_var_from_string(name, param_type, checks):
+def set_one_var_from_string(
+    name: Optional[str],
+    param_type: Optional[str],
+    checks: List
+) -> Optional[str]:
```
Comment (jameslamb) [inline helpers/parameter_generator.py:168]:
```suggestion
    name: str,
    param_type: str,
    checks: List[str]
) -> str:
```

File: helpers/parameter_generator.py [hunk 6]
```diff
@@ -189,7 +203,11 @@ def set_one_var_from_string(name, param_type, checks):
     return ret
 
 
-def gen_parameter_description(sections, descriptions, params_rst):
+def gen_parameter_description(
+    sections: List,
+    descriptions: List,
+    params_rst: Path
+):
```
Comment (jameslamb) [inline helpers/parameter_generator.py:210]:
```suggestion
def gen_parameter_description(
    sections: List[Tuple[str, int]],
    descriptions: List[List[Dict[str, Any]]],
    params_rst: Path
) -> None:
```

Functions that return nothing should be annotated with `-> None`

File: helpers/parameter_generator.py [hunk 7]
```diff
@@ -251,7 +269,10 @@ def gen_parameter_description(sections, descriptions, params_rst):
         new_params_file.write(after)
 
 
-def gen_parameter_code(config_hpp, config_out_cpp):
+def gen_parameter_code(
+    config_hpp: Path,
+    config_out_cpp: Path
+) -> Tuple[List, List]:
```
Comment (jameslamb) [inline helpers/parameter_generator.py:275]:
```suggestion
) -> Tuple[List[Tuple[str, int]], List[List[Dict[str, Any]]]]:
```

File: helpers/parameter_generator.py [hunk 8]
```diff
@@ -79,7 +82,9 @@ def get_parameter_infos(config_hpp):
     return keys, member_infos
 
 
-def get_names(infos):
+def get_names(
+    infos: List
+) -> List:
```
Comment (jameslamb) [inline helpers/parameter_generator.py:87]:
```suggestion
    infos: List[List[Dict[str, Any]]]
) -> List[str]:
```

File: helpers/parameter_generator.py [hunk 9]
```diff
@@ -7,9 +7,12 @@
 from the information in LightGBM/include/LightGBM/config.h file.
 """
 from pathlib import Path
+from typing import List, Optional, Tuple
```
Comment (jameslamb) [inline helpers/parameter_generator.py:10]:
```suggestion
from typing import Any, Dict, List, Optional, Tuple
```

Adding the additional imports needed based on other suggested changes in review comments.

File: helpers/parameter_generator.py [hunk 10]
```diff
@@ -99,7 +104,9 @@ def get_names(infos):
     return names
 
 
-def get_alias(infos):
+def get_alias(
+    infos: List
```
Comment (jameslamb) [inline helpers/parameter_generator.py:108]:
```suggestion
    infos: List[List[Tuple[Dict[str, Any]]]]
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your help! Please see the suggestions from my first review. You can accept them directly in the browser by clicking "add to batch" here on GitHub, then "commit suggestions".

These suggestions could be summarized by these three practices:

* when using types which can contain instances of other types, add type hints for those contents. For example, a list of tuples of 3 integers should be `List[Tuple[int, int, int]]`, not `List`
* only use `Optional` for something which can be `None`
* give functions which explicitly return `None` or implicitly return it by not having any `return` statements the return hint `-> None`

**Comment by sagnik1511** [comment]
Thanks, @jameslamb  for reviewing, I understood that the codes should be more specific and less coded. Thank You, sir.

**Review by jameslamb — APPROVED** [review]
Thanks very much!

I found some additional issues by running `mypy helpers/parameter_generator.py`.

```text
helpers/parameter_generator.py:31: error: Need type annotation for 'cur_info' (hint: "cur_info: Dict[<type>, <type>] = ...")
helpers/parameter_generator.py:33: error: Need type annotation for 'member_infos' (hint: "member_infos: List[<type>] = ...")
helpers/parameter_generator.py:126: error: Invalid tuple index type (actual type "str", expected type "Union[int, slice]")
helpers/parameter_generator.py:127: error: Invalid tuple index type (actual type "str", expected type "Union[int, slice]")
helpers/parameter_generator.py:291: error: Argument 1 to "get_names" has incompatible type "List[Dict[str, Any]]"; expected "List[List[Dict[str, Any]]]"
helpers/parameter_generator.py:292: error: Argument 1 to "get_alias" has incompatible type "List[Dict[str, Any]]"; expected "List[List[Tuple[Dict[str, Any]]]]"
helpers/parameter_generator.py:328: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
helpers/parameter_generator.py:329: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
helpers/parameter_generator.py:330: error: Need type annotation for 'checks' (hint: "checks: List[<type>] = ...")
helpers/parameter_generator.py:332: error: Incompatible types in assignment (expression has type "str", variable has type "List[Any]")
helpers/parameter_generator.py:332: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
helpers/parameter_generator.py:343: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
helpers/parameter_generator.py:344: error: Invalid index type "str" for "str"; expected type "Union[int, slice]"
helpers/parameter_generator.py:359: error: Incompatible return value type (got "Tuple[List[Tuple[str, int]], List[Dict[str, Any]]]", expected "Tuple[List[Tuple[str, int]], List[List[Dict[str, Any]]]]")
Found 14 errors in 1 file (checked 1 source file)
```

I pushed https://github.com/sagnik1511/LightGBM/commit/11ed4b3f0545948672bb2735875ba9cf64df6b91 to this branch to fix them.

**Comment by sagnik1511** [comment]
Thank You @jameslamb, for that.

**Comment by jameslamb** [comment]
> Can I try to update other files?

Sure, please open a new pull request if you'd like to update a new file. Please consider the suggestion in https://github.com/microsoft/LightGBM/pull/4474#pullrequestreview-707998470 when you do that.

File: helpers/parameter_generator.py [hunk 11]
```diff
@@ -7,9 +7,12 @@
 from the information in LightGBM/include/LightGBM/config.h file.
 """
 from pathlib import Path
+from typing import Any, Dict, List, Optional, Tuple
 
 
-def get_parameter_infos(config_hpp):
+def get_parameter_infos(
+    config_hpp: Path
+) -> Tuple[List[Tuple[str, int]], List[List[Dict[str, Any]]]]:
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:15]:
```suggestion
) -> Tuple[List[Tuple[str, int]], List[List[Dict[str, List]]]]:
```

File: helpers/parameter_generator.py [hunk 12]
```diff
@@ -25,9 +28,9 @@ def get_parameter_infos(config_hpp):
     is_inparameter = False
     cur_key = None
     key_lvl = 0
-    cur_info = {}
+    cur_info: Dict[str, Any] = {}
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:31]:
```suggestion
    cur_info: Dict[str, List] = {}
```

File: helpers/parameter_generator.py [hunk 13]
```diff
@@ -25,9 +28,9 @@ def get_parameter_infos(config_hpp):
...
     keys = []
-    member_infos = []
+    member_infos: List[List[Dict[str, Any]]] = []
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:33]:
```suggestion
    member_infos: List[List[Dict[str, List]]] = []
```

File: helpers/parameter_generator.py [hunk 14]
```diff
@@ -76,10 +79,13 @@ def get_parameter_infos(config_hpp):
                             cur_info["name"] = [tokens[1][:-1].strip()]
                     member_infos[-1].append(cur_info)
                     cur_info = {}
+
     return keys, member_infos
 
 
-def get_names(infos):
+def get_names(
+    infos: List[List[Dict[str, Any]]]
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:87]:
```suggestion
    infos: List[List[Dict[str, List]]]
```

File: helpers/parameter_generator.py [hunk 15]
```diff
@@ -99,7 +105,9 @@ def get_names(infos):
     return names
 
 
-def get_alias(infos):
+def get_alias(
+    infos: List[List[Dict[str, Any]]]
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:109]:
```suggestion
    infos: List[List[Dict[str, List]]]
```

File: helpers/parameter_generator.py [hunk 16]
```diff
@@ -189,7 +204,11 @@ def set_one_var_from_string(name, param_type, checks):
     return ret
 
 
-def gen_parameter_description(sections, descriptions, params_rst):
+def gen_parameter_description(
+    sections: List[Tuple[str, int]],
+    descriptions: List[List[Dict[str, Any]]],
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:209]:
```suggestion
    descriptions: List[List[Dict[str, List]]],
```

File: helpers/parameter_generator.py [hunk 17]
```diff
@@ -251,7 +270,10 @@ def gen_parameter_description(sections, descriptions, params_rst):
         new_params_file.write(after)
 
 
-def gen_parameter_code(config_hpp, config_out_cpp):
+def gen_parameter_code(
+    config_hpp: Path,
+    config_out_cpp: Path
+) -> Tuple[List[Tuple[str, int]], List[List[Dict[str, Any]]]]:
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:276]:
```suggestion
) -> Tuple[List[Tuple[str, int]], List[List[Dict[str, List]]]]:
```

**Review by StrikerRUS — COMMENTED** [review]
I think we can be even more precise and use `List` instead of `Any`:

File: helpers/parameter_generator.py (hunk 17 above)
Comment (sagnik1511) [reply helpers/parameter_generator.py:276]: Okay, I'm updating those.

**Review by StrikerRUS — APPROVED** [review]
LGTM, thank you!

**Comment by sagnik1511** [comment]
@StrikerRUS, thank you for letting me collaborate. Please, do merge this PR :)

**Comment by jameslamb** [comment]
Sorry for the delay in merging, thanks again for the help!

**Pull request merged**